content/posts/api-deployment-strategies.md

```diff
@@ -0,0 +1,341 @@
+---
+title: "API Deployment Strategies"
+date: 2023-09-15T09:01:51+05:30
+draft: true
+ShowToc: true
+TocOpen: true
+ShowRelatedContent: true
+summary: "Applying best practices to make new API deployments using the API gateway of choice, Apache APISIX."
+tags: ["apache apisix", "api gateway", "cloud-native"]
+categories: ["API Gateway"]
+cover:
+  image: "/images/api-deployment-strategies/shadow-banner.jpeg"
+  alt: "Shadow in a parking lot."
+  caption: "Deploying changes without affecting users is key to a good deployment strategy."
+  relative: false
+---
+
+Deploying changes to your APIs without impacting clients is vital in most modern software products. In this article, I will focus on some best practices and strategies you can adopt to streamline the deployment of your APIs using API gateways.
+
+Throughout this tutorial, we will use our API gateway of choice, [Apache APISIX](https://apisix.apache.org/), but you can translate these strategies to any other [API gateway](/posts/gateway-and-mesh/) or reverse proxy.
+
+## API Versions
+
+Yes, even that minor change can potentially break your client applications.
+
+API versioning is the most straightforward strategy to make changes to your APIs without disrupting your clients. By versioning, you can let your clients decide whether to upgrade to the new API version. They can also update their applications to the new API at their own pace.
+
+API gateways like APISIX can handle typical scenarios where your clients specify which version of the API to use in their requests. The version is typically specified in one of three ways:
+
+1. In URI path (`/v1/products` and `/v2/products`)
+2. In query parameters (`/products?version=v1` and `/products?version=v2`)
+3. In request headers (`/products -H 'version: v1'` and `/products -H 'version: v2'`)
```

Comment:
That's the naive version of Request Header versioning. Most REST-minded designers will use `Accept` header, _e.g._, `Accept: application/json; version=1` or even better `Accept: application/vnd.myapplication.user.v1+json`. It would be nice to use one of the latter as an example

cf. https://blog.allegro.tech/2015/01/Content-headers-or-how-to-version-api.html#version-in-header